modules/function/src/main/java/com/opengamma/strata/function/calculation/DefaultLookupRatesProvider.java

```diff
@@ -0,0 +1,280 @@
+
+/**
+ * A rates provider based on a rates lookup.
+ * <p>
+ * This uses a {@link DefaultRatesMarketDataLookup} to provide a view on {@link MarketData}.
+ */
+@BeanDefinition(style = "light")
+final class DefaultLookupRatesProvider
+    extends AbstractRatesProvider
+    implements RatesProvider, ImmutableBean {
+
+  /**
+   * The lookup.
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final DefaultRatesMarketDataLookup lookup;
+  /**
+   * The market data.
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final MarketData marketData;
+  /**
+   * The FX rate provider.
+   */
+  private final FxRateProvider fxRateProvider;  // derived
+
+  //-------------------------------------------------------------------------
+  /**
+   * Obtains an instance based on a lookup and market data.
+   * <p>
+   * The lookup provides the mapping from currency to discount curve, and from
+   * index to forward curve. The curves are in the market data.
+   *
+   * @param lookup  the lookup
+   * @param marketData  the market data
+   * @return the rates provider
+   */
+  public static DefaultLookupRatesProvider of(DefaultRatesMarketDataLookup lookup, MarketData marketData) {
+    return new DefaultLookupRatesProvider(lookup, marketData);
+  }
+
+  @ImmutableConstructor
+  private DefaultLookupRatesProvider(DefaultRatesMarketDataLookup lookup, MarketData marketData) {
+    this.lookup = ArgChecker.notNull(lookup, "lookup");
+    this.marketData = ArgChecker.notNull(marketData, "marketData");
+    this.fxRateProvider = MarketDataFxRateProvider.of(marketData);
+  }
+
+  //-------------------------------------------------------------------------
+  @Override
+  public LocalDate getValuationDate() {
+    return marketData.getValuationDate();
+  }
+
+  //-------------------------------------------------------------------------
+  @Override
+  public <T> T data(MarketDataKey<T> key) {
+    return marketData.getValue(key);
+  }
+
+  //-------------------------------------------------------------------------
+  @Override
+  public Optional<Curve> findCurve(CurveName name) {
+    return Stream.concat(lookup.getDiscountCurves().values().stream(), lookup.getForwardCurves().values().stream())
+        .filter(id -> id.getCurveName().equals(name))
+        .findFirst()
+        .flatMap(id -> marketData.findValue(id));
+  }
+
+  @Override
+  public LocalDateDoubleTimeSeries timeSeries(Index index) {
+    return marketData.getTimeSeries(IndexRateKey.of(index));
+  }
+
+  //-------------------------------------------------------------------------
+  @Override
+  public double fxRate(Currency baseCurrency, Currency counterCurrency) {
+    return fxRateProvider.fxRate(baseCurrency, counterCurrency);
+  }
+
+  //-------------------------------------------------------------------------
+  @Override
+  public DiscountFactors discountFactors(Currency currency) {
+    SimpleCurveId curveId = lookup.getDiscountCurves().get(currency);
```

Comment:
This relies on referring to the implementation type. That's a bit of a smell. Do we need to add something to the interface?